### PR TITLE
Bug fix for manipulating submodules in aspects

### PR DESCRIPTION
**Pull request by jdeters**
**Related issue**: 
This fixes https://github.com/freechipsproject/chisel3/issues/1537

**Type of change**: bug report

**Impact**: API modification

**Development Phase**: implementation

**Release Notes**

Adds functionality for modules' parents to be translated into their corresponding aspect so they can be manipulated in an aspect that injects into the parent module.

### CODE REVIEW

**Comment by azidar** [comment]
Awesome!!! I'll need to take a closer look, but could you please provide a unit test demonstrating the bug? Thanks!

**Comment by jdeters** [comment]
I added a unit test into the InjectionSpec that demonstrates this bug. 

As for the logic, I personally don't think that it's a _generalization_ of current functionality, but actually a more _specific_ case. Left/right/source/sink currently describe the _current_ module being manipulated. However, when we check the parent for legality of connection, we also need to check if that parent has an attached aspect. The kicker here is that it only really matters if we're currently elaborating an aspect module. Since this happens so infrequently, I had the normal parent check logic short circuit. To be fair, *_aspect should probably be named *_parentAspect.

**Comment by jdeters** [comment]
I put some more thought into it and I agree that the logic can be made simpler. I just refactored the changes to mostly reflect what @ducky64 was asking for.

File: core/src/main/scala/chisel3/internal/Builder.scala [hunk 1]
```diff
@@ -427,6 +427,35 @@ private[chisel3] object Builder {
     case Some(dynamicContext) => dynamicContext.aspectModule.get(module)
     case _ => None
   }
+  /**determines if the parent module may have an attached aspect
+  if there isn't a parent, return None
+  if the parent doesn't have and attached aspect, return the parent
+  */
+  def parentAspectModule(module: BaseModule): Option[BaseModule] = {
```
[the remark this answers is not in the document]
Comment (jdeters) [reply core/src/main/scala/chisel3/internal/Builder.scala:434]: This is my bad, this should have been removed. I used this in the previous method for finding the correct parent.

**Comment by jdeters** [comment]
> (Possibly) random idea: does it make sense to change `Module._parent` (or make a new interface, eg `Module.getParent`) that would return the parent considering aspects? Put another way, is there any reason we want the current (narrow) `Module._parent` instead of the more expansive `Builder.retrieveParent`?

I also considered this, but didn't want to suggest a change that fundamental (as I've only been working with Scala and Chisel for a few weeks now). However, one reason I would keep `Builder.retrieveParent` is that the proper parent does depend on the _context_ of the module being elaborated and this is explicit about providing that context.

**Comment by ekiwi-sifive** [comment]
It seems like somehow this commit broke some Chisel tests. @azidar

**Comment by azidar** [comment]
Hmm, let's revert it.

**Comment by seldridge** [comment]
Hang on. I'm pretty sure this is the CustomFileEmission in FIRRTL. I need https://github.com/freechipsproject/chisel3/pull/1405 to get merged for tests to stop breaking.

**Comment by ekiwi-sifive** [comment]
> Hang on. I'm pretty sure this is the CustomFileEmission in FIRRTL.

Oh, yeah, I forgot about that.

The failing tests are:

```
[error] Failed tests:
[error] 	chiselTests.DontTouchSpec
[error] 	chiselTests.stage.ChiselStageSpec
[error] 	chiselTests.experimental.verification.VerificationSpec
[error] 	chiselTests.ResetSpec
[error] 	chiselTests.MemorySpec
[error] 	chiselTests.DedupSpec
[error] 	chiselTests.MuxLookupExhaustiveSpec
[error] 	chiselTests.CompatibiltyInteroperabilitySpec
```

**Comment by ekiwi-sifive** [comment]
Thanks @seldridge ! In the end this PR wasn't responsible for the failure.

**Pull request merged**